compiler/rustc_errors/src/lib.rs

```diff
@@ -807,6 +809,23 @@ impl DiagCtxt {
         Some(Diag::new_diagnostic(self, diag))
     }
 
+    /// Steals a previously stashed error with the given `Span` and
+    /// [`StashKey`] as the key, and cancels it if found.
+    /// Panics if the found diagnostic's level isn't `Level::Error`.
+    pub fn steal_err(&self, span: Span, key: StashKey, _: ErrorGuaranteed) -> bool {
```

Comment:
Not sure if this makes sense; I needed something like that in `parse_pat_tuple_or_parens()` to avoid emitting too many diagnostics since each parenthesis could have a trailing expression:

```
(1 + 2) * 3
   ~~~  ~~~
```
```
(1 + 2) * 3
 ~~~~~

(1 + 2) * 3
~~~~~~~~~~~
```

Reply:
@nnethercote Does this look okay to you?

Reply:
Stashed diagnostics are tricky. Can you use one of the existing methods? Either `try_steal_modify_and_emit_err` or `try_steal_replace_and_emit_err`? I would strongly prefer that.

Reply:
Welp, it's unsound:
```rs
let guar = dcx.create_err(err).stash(span, key).unwrap();
dcx.steal_err(span, key, guar);
guar // `ErrorGuaranteed`, even tough it was cancelled
```
I'll look for an alternative.

Reply:
I'm confused... what is unsound?

More generally, I'm very reluctant to muck around with the stashing API, adding new things to it.

Reply:
Ah mb, I thought `stash_diagnostic` created an `ErrorGuaranteed` out of thin air, but it actually creates a delayed bug.

Anyway, I'll look for a way that doesn't require modifying the stashing API.

Reply:
[This comment](https://github.com/rust-lang/rust/blob/9337f7afa6fda07e60d6aa5ee88e692878446782/compiler/rustc_errors/src/lib.rs#L782-L802) explains what `stash_diagnostic` does.